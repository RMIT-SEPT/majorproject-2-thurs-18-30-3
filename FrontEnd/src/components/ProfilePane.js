import React from 'react'
import {Link} from 'react-router-dom'
import AuthService from '../services/auth.service'

import '../containers/App.css'

//Displays a user profile in a modal view

function ProfilePane({close, profile, reload, update}) {
  const [isEditing, setIsEditing] = React.useState(false)
  const [displayFirstName, setDisplayFirstName] = React.useState(profile.firstName)
  const [displayLastName, setDisplayLastName] = React.useState(profile.lastName)
  const [displayEmail, setDisplayEmail] = React.useState(profile.email)
  const [displayPhone, setDisplayPhone] = React.useState(profile.mobileNum)
  const [displayAddress, setDisplayAddress] = React.useState(profile.address)

  //Fires when clicking the cancel button - reinitialises profile data.
  const cancel = () => {
    setIsEditing(!isEditing)
    setDisplayFirstName(profile.firstName)
    setDisplayLastName(profile.lastName)
    setDisplayEmail(profile.email)
    setDisplayPhone(profile.mobileNum)
    setDisplayAddress(profile.address)
  }

  //Fires when clicking the save button - commits profile data.
  const save = () => {
    setIsEditing(!isEditing)
    update(displayEmail, displayFirstName, displayLastName, displayPhone, displayAddress)
    reload()
  }

  return (
    <div className="modal-wrapper">
      <main className="modalPane">
        <div className="paneHeader">
          <button className="closeButton" onClick={() => close()}>
            &times;
          </button>

          {/*TODO:Display image*/}
          <div className="book-bubble" />

          {/*Data is displayed in editable fields - this button enables/disables editing*/}
          <button className="editButton" onClick={() => cancel()}>
            {isEditing ? 'cancel' : 'edit'}
          </button>
        </div>
        <div className="paneBody">
          <label htmlFor="fname">name</label>
          <input
            id="fname"
            name="fname"
            className="paneInput"
            disabled={!isEditing}
            value={displayFirstName}
            onChange={(event) => {
              setDisplayFirstName(event.target.value)
            }}
          />

          <label htmlFor="lname">surname</label>
          <input
            id="lname"
            name="lname"
            className="paneInput"
            disabled={!isEditing}
            value={displayLastName}
            onChange={(event) => {
              setDisplayLastName(event.target.value)
            }}
          />

          <label htmlFor="email">email</label>
          <input
            id="email"
            name="email"
            className="paneInput"
            disabled={!isEditing}
            value={displayEmail}
            onChange={(event) => {
              setDisplayEmail(event.target.value)
            }}
          />

          <label htmlFor="phNum">phone number</label>
          <input
            id="phNum"
            name="phNum"
            className="paneInput"
            disabled={!isEditing}
            value={displayPhone}
            onChange={(event) => {
              setDisplayPhone(event.target.value)
            }}
          />

          <label htmlFor="phNum">address</label>
          <input
            id="address"
            name="address"
            className="paneInput"
            disabled={!isEditing}
            value={displayAddress}
            onChange={(event) => {
              setDisplayAddress(event.target.value)
            }}
          />
          <div className="button-set">
            {isEditing ? (
              <button className="actButton" onClick={() => save()}>
                save
              </button>
            ) : (
              <>
                <button className="actButton">bookings</button>
                <Link to="/service">
                  {/*Logout button also closes the pane and reloads the page*/}
                  <button
                    className="actButton"
                    onClick={() => {
                      AuthService.logout()
                      close()
                      window.location.reload()
                    }}>
                    log out
                  </button>
                </Link>
              </>
            )}
          </div>
        </div>
      </main>
    </div>
  )
}
export default ProfilePane
