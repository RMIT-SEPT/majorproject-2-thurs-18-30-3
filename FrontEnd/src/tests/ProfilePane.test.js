import React from 'react';

import ProfilePane from '../components/ProfilePane';
import { StaticRouter } from 'react-router';
import renderer from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';
const dummyProfile = {firstName: 'test', lastName: 'tester', email: 'test@wiz', mobileNum:'0000', userType: 'customer', address: '001 happy test avenue'};
const closeFuncMock = jest.fn();
const changeFuncMock = jest.fn();
const reloadFuncMock = jest.fn();
const updateFuncMock = jest.fn();

const setup = () => {
    return render(
        <StaticRouter>
            <ProfilePane close={closeFuncMock} change = {changeFuncMock} profile={dummyProfile} reload = {reloadFuncMock} update = {updateFuncMock}/>
        </StaticRouter>
    )
  }

  //snapshot test
test('Test Pane Renders', () => {
    const pane = renderer.create(
        <StaticRouter>
            <ProfilePane close={closeFuncMock} change = {changeFuncMock} profile={dummyProfile} reload = {reloadFuncMock} update = {updateFuncMock}/>
        </StaticRouter>,
      ).toJSON();
      expect(pane).toMatchSnapshot();
});

test('Test Pane Contents', () => {
    const container = setup();
    
    expect(container.getAllByRole('button').length).toBe(4);
    expect(container.getAllByRole('textbox').length).toBe(5);
});

//Make sure input fields are in correct order
test('Test Pane Data', () => {
    const container = setup();
    
    const fname= container.getAllByRole('textbox')[0];
    const lname= container.getAllByRole('textbox')[1];
    const email= container.getAllByRole('textbox')[2];
    const phone= container.getAllByRole('textbox')[3];
    const address= container.getAllByRole('textbox')[4];

    expect(fname.value).toBe('test');
    expect(lname.value).toBe('tester');
    expect(email.value).toBe('test@wiz');
    expect(phone.value).toBe('0000');
    expect(address.value).toBe('001 happy test avenue');

});

//Check that buttons have correct functions
test('Test Pane Functions', () => {
    const container = setup();
    
    const closeBtn= container.getAllByRole('button')[0];
    const editBtn= container.getAllByRole('button')[1];
    const bookingsBtn= container.getAllByRole('button')[2];


    fireEvent.click(closeBtn);
    expect(closeFuncMock).toHaveBeenCalled();

    fireEvent.click(bookingsBtn);
    expect(changeFuncMock).toHaveBeenCalled();

    fireEvent.click(editBtn);
    const saveBtn= container.getAllByRole('button')[2];
    expect(saveBtn).toHaveTextContent('save');
});

test('Test Pane Validation', () => {
    const container = setup();

    const editBtn= container.getAllByRole('button')[1];

    const fname= container.getAllByRole('textbox')[0];
    const lname= container.getAllByRole('textbox')[1];
    const email= container.getAllByRole('textbox')[2];
    const phone= container.getAllByRole('textbox')[3];
    const address= container.getAllByRole('textbox')[4];

    fireEvent.click(editBtn);
    const saveBtn= container.getAllByRole('button')[2];
    
    fireEvent.click(saveBtn);
    expect(updateFuncMock).toBeCalledTimes(0);

    fireEvent.change(fname, {target: {value: 'new fname'}});
    fireEvent.change(lname, {target: {value: 'new lname'}});
    fireEvent.change(email, {target: {value: 'new email'}});
    fireEvent.change(phone, {target: {value: 'new phone'}});
    fireEvent.change(address, {target: {value: 'new address'}});

    fireEvent.click(saveBtn);
    expect(updateFuncMock).toBeCalledTimes(0);

    fireEvent.change(email, {target: {value: 'test@test.com'}});
    fireEvent.change(phone, {target: {value: '111111'}});

    fireEvent.click(saveBtn);
    expect(updateFuncMock).toHaveBeenCalled();

    expect(fname.value).toBe('new fname');
    expect(lname.value).toBe('new lname');
    expect(email.value).toBe('test@test.com');
    expect(phone.value).toBe('111111');
    expect(address.value).toBe('new address');
});