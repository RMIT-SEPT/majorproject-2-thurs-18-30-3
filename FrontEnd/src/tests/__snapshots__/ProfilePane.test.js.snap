// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Pane Renders 1`] = `
<div
  className="modal-wrapper"
>
  <main
    className="modalPane"
  >
    <div
      className="paneHeader"
    >
      <button
        className="closeButton"
        onClick={[Function]}
      >
        <i
          className="material-icons md-32"
        >
           close 
        </i>
      </button>
      <h2>
        <br />
        Your Profile
      </h2>
      <button
        className="editButton"
        onClick={[Function]}
      >
        edit
      </button>
    </div>
    <div
      className="paneBody"
    >
      <label
        htmlFor="fname"
      >
        name
      </label>
      <input
        className="paneInput"
        disabled={true}
        id="fname"
        name="fname"
        onChange={[Function]}
        value="test"
      />
      <label
        htmlFor="lname"
      >
        surname
      </label>
      <input
        className="paneInput"
        disabled={true}
        id="lname"
        name="lname"
        onChange={[Function]}
        value="tester"
      />
      <label
        htmlFor="email"
      >
        email
      </label>
      <input
        className="paneInput"
        disabled={true}
        id="email"
        name="email"
        onChange={[Function]}
        value="test@wiz"
      />
      <label
        htmlFor="phNum"
      >
        phone number
      </label>
      <input
        className="paneInput"
        disabled={true}
        id="phNum"
        name="phNum"
        onChange={[Function]}
        pattern="[A-Za-z]{3}"
        value="0000"
      />
      <label
        htmlFor="address"
      >
        address
      </label>
      <input
        className="paneInput"
        disabled={true}
        id="address"
        name="address"
        onChange={[Function]}
        value="001 happy test avenue"
      />
      <div
        className="button-set"
      >
        <button
          className="actButton"
          onClick={[Function]}
        >
          bookings
        </button>
        <a
          href="/service"
          onClick={[Function]}
        >
          <button
            className="actButton"
            onClick={[Function]}
          >
            log out
          </button>
        </a>
      </div>
    </div>
  </main>
</div>
`;
